File: go/cardinality_test.go
// SPDX-FileCopyrightText: Copyright (C) 2026 provide.io llc
// SPDX-License-Identifier: Apache-2.0

package telemetry

import (
	"fmt"
	"sync"
	"testing"
	"time"
)

func TestGuardAttributes_NoLimit_PassThrough(t *testing.T) {
	_resetCardinalityLimits()
	t.Cleanup(_resetCardinalityLimits)

	input := map[string]string{"env": "production", "region": "us-east-1"}
	got := GuardAttributes(input)

	if got["env"] != "production" {
		t.Errorf("env: want %q, got %q", "production", got["env"])
	}
	if got["region"] != "us-east-1" {
		t.Errorf("region: want %q, got %q", "us-east-1", got["region"])
	}
}

func TestGuardAttributes_SeenValue_PassThrough(t *testing.T) {
	_resetCardinalityLimits()
	t.Cleanup(_resetCardinalityLimits)

	SetCardinalityLimit("user_id", CardinalityLimit{MaxValues: 3, TTLSeconds: 60})

	// First call: adds value to cache.
	got1 := GuardAttributes(map[string]string{"user_id": _testAlice})
	if got1["user_id"] != _testAlice {
		t.Fatalf("first call: want %q, got %q", _testAlice, got1["user_id"])
	}

	// Second call: same value is already cached — must not overflow.
	got2 := GuardAttributes(map[string]string{"user_id": _testAlice})
	if got2["user_id"] != _testAlice {
		t.Errorf("second call (seen): want %q, got %q", _testAlice, got2["user_id"])
	}
}

func TestGuardAttributes_UnderLimit_Added(t *testing.T) {
	_resetCardinalityLimits()
	t.Cleanup(_resetCardinalityLimits)

	SetCardinalityLimit("status", CardinalityLimit{MaxValues: 5, TTLSeconds: 60})

	for _, v := range []string{"ok", "error", "timeout"} {
		got := GuardAttributes(map[string]string{"status": v})
		if got["status"] != v {
			t.Errorf("status=%q: want %q, got %q", v, v, got["status"])
		}
	}
}

func TestGuardAttributes_AtLimit_Overflow(t *testing.T) {
	_resetCardinalityLimits()
	t.Cleanup(_resetCardinalityLimits)

	SetCardinalityLimit("tenant_id", CardinalityLimit{MaxValues: 2, TTLSeconds: 60})

	// Fill to capacity.
	r1 := GuardAttributes(map[string]string{"tenant_id": "t1"})
	if r1["tenant_id"] != "t1" {
		t.Fatalf("slot 1: want %q, got %q", "t1", r1["tenant_id"])
	}
	r2 := GuardAttributes(map[string]string{"tenant_id": "t2"})
	if r2["tenant_id"] != "t2" {
		t.Fatalf("slot 2: want %q, got %q", "t2", r2["tenant_id"])
	}

	// New value when at capacity.
	r3 := GuardAttributes(map[string]string{"tenant_id": "t3"})
	if r3["tenant_id"] != _overflowValue {
		t.Errorf("over limit: want %q, got %q", _overflowValue, r3["tenant_id"])
	}
}

func TestGuardAttributes_TTLExpiry_SlotFreed(t *testing.T) {
	_resetCardinalityLimits()
	t.Cleanup(_resetCardinalityLimits)

	SetCardinalityLimit("session", CardinalityLimit{MaxValues: 1, TTLSeconds: 0.01}) // 10ms TTL

	// Fill the single slot.
	r1 := GuardAttributes(map[string]string{"session": "s1"})
	if r1["session"] != "s1" {
		t.Fatalf("slot 1: want %q, got %q", "s1", r1["session"])
	}

	// New value immediately overflows.
	r2 := GuardAttributes(map[string]string{"session": "s2"})
	if r2["session"] != _overflowValue {
		t.Fatalf("before TTL: want %q, got %q", _overflowValue, r2["session"])
	}

	// Wait for TTL to expire.
	time.Sleep(20 * time.Millisecond)

	// After expiry the slot is free — new value should be accepted.
	r3 := GuardAttributes(map[string]string{"session": "s2"})
	if r3["session"] != "s2" {
		t.Errorf("after TTL: want %q, got %q", "s2", r3["session"])
	}
}

func TestSetGetCardinalityLimit_RoundTrip(t *testing.T) {
	_resetCardinalityLimits()
	t.Cleanup(_resetCardinalityLimits)

	want := CardinalityLimit{MaxValues: 42, TTLSeconds: 3.14}
	SetCardinalityLimit("foo", want)
	got := GetCardinalityLimit("foo")

	if got.MaxValues != want.MaxValues {
		t.Errorf("MaxValues: want %d, got %d", want.MaxValues, got.MaxValues)
	}
	if got.TTLSeconds != want.TTLSeconds {
		t.Errorf("TTLSeconds: want %v, got %v", want.TTLSeconds, got.TTLSeconds)
	}
}

func TestGetCardinalityLimit_Unknown_ZeroValue(t *testing.T) {
	_resetCardinalityLimits()
	t.Cleanup(_resetCardinalityLimits)

	got := GetCardinalityLimit("nonexistent")
	if got.MaxValues != 0 || got.TTLSeconds != 0 {
		t.Errorf("unknown key: want zero CardinalityLimit, got %+v", got)
	}
}

func TestGuardAttributes_MultipleKeys_Independent(t *testing.T) {
	_resetCardinalityLimits()
	t.Cleanup(_resetCardinalityLimits)

	SetCardinalityLimit("color", CardinalityLimit{MaxValues: 1, TTLSeconds: 60})
	SetCardinalityLimit("size", CardinalityLimit{MaxValues: 1, TTLSeconds: 60})

	// Fill both keys.
	GuardAttributes(map[string]string{"color": "red", "size": "large"})

	// New values for both should overflow independently.
	result := GuardAttributes(map[string]string{"color": "blue", "size": "small"})

	if result["color"] != _overflowValue {
		t.Errorf("color: want %q, got %q", _overflowValue, result["color"])
	}
	if result["size"] != _overflowValue {
		t.Errorf("size: want %q, got %q", _overflowValue, result["size"])
	}
}

func TestGuardAttributes_Concurrent(t *testing.T) {
	_resetCardinalityLimits()
	t.Cleanup(_resetCardinalityLimits)

	SetCardinalityLimit("ckey", CardinalityLimit{MaxValues: 10, TTLSeconds: 60})

	const goroutines = 50
	const iterations = 100

	var wg sync.WaitGroup
	wg.Add(goroutines)
	for i := 0; i < goroutines; i++ {
		go func(id int) {
			defer wg.Done()
			for j := 0; j < iterations; j++ {
				val := fmt.Sprintf("v%d", j%20) // 20 distinct values, limit=10 → half overflow
				result := GuardAttributes(map[string]string{"ckey": val})
				v := result["ckey"]
				if v != val && v != _overflowValue {
					t.Errorf("unexpected value %q for input %q", v, val)
				}
			}
		}(i)
	}
	wg.Wait()
}

File: rust/tests/cardinality_test.rs
// SPDX-FileCopyrightText: Copyright (C) 2026 provide.io llc
// SPDX-License-Identifier: Apache-2.0
// SPDX-Comment: Part of provide-telemetry.
//

use std::collections::HashMap;
use std::thread;
use std::time::Duration;

use provide_telemetry::cardinality::OVERFLOW_VALUE;
use provide_telemetry::testing::{acquire_test_state_lock, reset_telemetry_state};
use provide_telemetry::{CardinalityLimit, guard_attributes, register_cardinality_limit};

#[test]
fn cardinality_test_overflows_after_max_distinct_values_seen_for_key() {
    let _guard = acquire_test_state_lock();
    reset_telemetry_state();

    register_cardinality_limit(
        "user.id",
        CardinalityLimit {
            max_values: 1,
            ttl_seconds: 60.0,
        },
    );

    let first = guard_attributes(HashMap::from([("user.id".to_string(), "u-1".to_string())]));
    assert_eq!(first.get("user.id").map(String::as_str), Some("u-1"));

    let second = guard_attributes(HashMap::from([("user.id".to_string(), "u-2".to_string())]));
    assert_eq!(
        second.get("user.id").map(String::as_str),
        Some(OVERFLOW_VALUE)
    );
}

#[test]
fn cardinality_test_reuses_existing_seen_value_without_overflow() {
    let _guard = acquire_test_state_lock();
    reset_telemetry_state();

    register_cardinality_limit(
        "tenant.id",
        CardinalityLimit {
            max_values: 1,
            ttl_seconds: 60.0,
        },
    );

    let first = guard_attributes(HashMap::from([(
        "tenant.id".to_string(),
        "tenant-a".to_string(),
    )]));
    assert_eq!(first.get("tenant.id").map(String::as_str), Some("tenant-a"));

    let second = guard_attributes(HashMap::from([(
        "tenant.id".to_string(),
        "tenant-a".to_string(),
    )]));
    assert_eq!(
        second.get("tenant.id").map(String::as_str),
        Some("tenant-a")
    );
}

#[test]
fn cardinality_test_guard_attributes_returns_expected_keys() {
    let _guard = acquire_test_state_lock();
    reset_telemetry_state();

    register_cardinality_limit(
        "env",
        CardinalityLimit {
            max_values: 2,
            ttl_seconds: 60.0,
        },
    );

    let guarded = guard_attributes(HashMap::from([
        ("env".to_string(), "prod".to_string()),
        ("region".to_string(), "us-east".to_string()),
    ]));

    assert_eq!(guarded.len(), 2);
    assert_eq!(guarded.get("env").map(String::as_str), Some("prod"));
    assert_eq!(guarded.get("region").map(String::as_str), Some("us-east"));
    assert!(!guarded.contains_key("xyzzy"));
}

#[test]
fn cardinality_test_ttl_expiry_reopens_capacity_for_new_value() {
    let _guard = acquire_test_state_lock();
    reset_telemetry_state();

    register_cardinality_limit(
        "session.id",
        CardinalityLimit {
            max_values: 1,
            ttl_seconds: 1.0,
        },
    );

    let first = guard_attributes(HashMap::from([(
        "session.id".to_string(),
        "s-1".to_string(),
    )]));
    assert_eq!(first.get("session.id").map(String::as_str), Some("s-1"));

    let overflow = guard_attributes(HashMap::from([(
        "session.id".to_string(),
        "s-2".to_string(),
    )]));
    assert_eq!(
        overflow.get("session.id").map(String::as_str),
        Some(OVERFLOW_VALUE)
    );

    thread::sleep(Duration::from_millis(5100));

    let after_ttl = guard_attributes(HashMap::from([(
        "session.id".to_string(),
        "s-3".to_string(),
    )]));
    assert_eq!(after_ttl.get("session.id").map(String::as_str), Some("s-3"));
}
